{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import h5py\n",
    "import numpy as np\n",
    "import hdf5plugin  # Ensure this is imported to enable the plugins\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "def nucleotide_to_index(seq, encode_spec=None):\n",
    "    \"\"\"\n",
    "    Convert a DNA sequence to integer indices.\n",
    "    \n",
    "    Parameters:\n",
    "    seq (str): A string representing a DNA sequence.\n",
    "    encode_spec (dict, optional): Encoding specification for nucleotides. Defaults to None.\n",
    "    \n",
    "    Returns:\n",
    "    np.array: An array of integers representing the indices of nucleotides.\n",
    "    \"\"\"\n",
    "    if encode_spec is None:\n",
    "        encode_spec = {'A': 0, 'C': 1, 'G': 2, 'T': 3, 'N': 4}\n",
    "    return np.array([encode_spec.get(nuc, 4) for nuc in seq], dtype=np.int8)\n",
    "\n",
    "def bitpack_indices(indices):\n",
    "    \"\"\"\n",
    "    Pack nucleotide indices into a 3-bit representation to include `N`.\n",
    "\n",
    "    Parameters:\n",
    "    indices (np.array): Array of nucleotide indices.\n",
    "\n",
    "    Returns:\n",
    "    np.array: Packed array of indices in 3-bit representation.\n",
    "    \"\"\"\n",
    "    packed = np.packbits(indices.reshape(-1, 2), axis=-1, bitorder='little')\n",
    "    return packed\n",
    "\n",
    "\n",
    "def index_to_onehot(indices, encode_spec=None):\n",
    "    \"\"\"\n",
    "    Convert nucleotide indices to one-hot encoding.\n",
    "    \n",
    "    Parameters:\n",
    "    indices (np.array): Array of nucleotide indices.\n",
    "    encode_spec (dict, optional): Encoding specification for nucleotides. Defaults to None.\n",
    "    \n",
    "    Returns:\n",
    "    np.array: One-hot encoded representation of the indices.\n",
    "    \"\"\"\n",
    "    if encode_spec is None:\n",
    "        encode_spec = {'A': 0, 'C': 1, 'G': 2, 'T': 3, 'N': 4}\n",
    "    num_classes = len(encode_spec)\n",
    "    return np.eye(num_classes)[indices]\n",
    "    \n",
    "def unpack_bits(packed_data):\n",
    "    \"\"\"\n",
    "    Unpack 3-bit packed data back to nucleotide indices.\n",
    "    \n",
    "    Parameters:\n",
    "    packed_data (np.array): Packed array of nucleotide indices.\n",
    "    \n",
    "    Returns:\n",
    "    np.array: Unpacked array of nucleotide indices.\n",
    "    \"\"\"\n",
    "    unpacked = np.unpackbits(packed_data, axis=-1, bitorder='little').reshape(-1, 2)\n",
    "    return unpacked\n",
    "\n",
    "\n",
    "def parse_encode_dict(encode_spec):\n",
    "    \"\"\"\n",
    "    Parse encoding specification into a dictionary.\n",
    "    \n",
    "    Parameters:\n",
    "    encode_spec (str, list, or dict): Encoding specification.\n",
    "    \n",
    "    Returns:\n",
    "    dict: Parsed encoding specification.\n",
    "    \"\"\"\n",
    "    if not encode_spec:\n",
    "        return {\"A\": 0, \"C\": 1, \"G\": 2, \"T\": 3, \"N\": 4}\n",
    "    elif isinstance(encode_spec, (list, tuple, str)):\n",
    "        return {base: i for i, base in enumerate(encode_spec)}\n",
    "    elif isinstance(encode_spec, dict):\n",
    "        return encode_spec\n",
    "    else:\n",
    "        raise TypeError(\"Please input as dict, list or string!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def read_and_verify_h5(file_path):\n",
    "    with h5py.File(file_path, 'r') as h5_file:\n",
    "        for donor in h5_file.keys():\n",
    "            donor_group = h5_file[donor]\n",
    "            for chrom in donor_group.keys():\n",
    "                print(f\"Verifying data for Donor: {donor}, Chromosome: {chrom}\")\n",
    "\n",
    "                # Read SNP data and genotypes\n",
    "                snp_data = donor_group[chrom]['snp_data'][:]\n",
    "                # packed_genotypes = donor_group[chrom]['genotypes'][:]\n",
    "                \n",
    "                # # Unpack genotypes\n",
    "                # unpacked_genotypes = np.unpackbits(packed_genotypes, axis=-1, bitorder='little').reshape(-1, 2)\n",
    "\n",
    "                # # Convert indices back to nucleotides\n",
    "                # ref_indices = snp_data['REF_IDX']\n",
    "                # alt_indices = snp_data['ALT_IDX']\n",
    "                # ref_nucleotides = ''.join([list(nucleotide_to_index.keys())[i] for i in ref_indices])\n",
    "                # alt_nucleotides = ''.join([list(nucleotide_to_index.keys())[i] for i in alt_indices])\n",
    "                \n",
    "                # Print data\n",
    "                print(f\"SNP Data: {snp_data}\")\n",
    "                # print(f\"Unpacked Genotypes: {unpacked_genotypes}\")\n",
    "                # print(f\"Reference Nucleotides: {ref_nucleotides}\")\n",
    "                # print(f\"Alternative Nucleotides: {alt_nucleotides}\")\n",
    "                print(\"\\n\")\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    file_path = '/iblm/netapp/data4/jjaureguy/vcf_cuda/HaploHyped-VarAwareML/tests/out/ipscs.h5'\n",
    "    read_and_verify_h5(file_path)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "haplo_test",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.19"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
